libnavui-maneuver/src/main/java/com/mapbox/navigation/ui/maneuver/RoadShieldDownloader.kt

```diff
@@ -0,0 +1,74 @@
+package com.mapbox.navigation.ui.maneuver
+
+import com.mapbox.common.HttpRequest
+import com.mapbox.common.core.module.CommonSingletonModuleProvider
+import com.mapbox.navigation.ui.maneuver.model.OnRoadShieldDownload
+import com.mapbox.navigation.utils.internal.ThreadController
+import kotlinx.coroutines.withContext
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+internal object RoadShieldDownloader {
+
+    private const val CODE_200 = 200L
+    private const val CODE_401 = 401L
+    private const val CODE_404 = 404L
+
+    suspend fun downloadImage(request: HttpRequest): OnRoadShieldDownload =
+        withContext(ThreadController.IODispatcher) {
+            suspendCoroutine { continuation ->
+                CommonSingletonModuleProvider.httpServiceInstance.request(request) { response ->
```

Comment:
It's not a blocking call, but if you look at the function `downloadImage` and backtrack it, you will see that it is ultimately being called from the main thread. It's fine for this operation to happen async, but I need to wait to get results before I process something else, because I need the route shield to construct my object. Also, since this code is in a separate file, using a `suspendible coroutine` helps to get back the result, otherwise I would have to use some sort of callback mechanism to get the results.

Reply:
I agree with this kind of implementation. I think it belongs on the IO thread because it's an http call and by wrapping the call in a suspendedCoroutine it means that a kotlin caller can use the method without a callback.

Reply:
In my opinion, this is an unnecessary thread switch. `httpServiceInstance.request` already executes asynchronously and uses the callback to get back to us with a result. I don't mind us using a `suspendCoroutine`, it's a thread switch that's unnecessary.